# **Pride and Joy**
### *An investigation of mental health correlates in LGBQ+ people*
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | |
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|Emily K. Sanders| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | |Capstone Project|
|DSB-318| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | |June 13, 2024|
---

In [1]:
#is there any way to space that text out so that it fits the edges of whatever screen it's on?

| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | |
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|Emily K. Sanders| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | |Capstone Project|
|DSB-318| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | |June 13, 2024|
---

## Prior Notebooks Summary

In the previous notebook, I 

- **gave a problem statement!**
- **listed deliverables!**
- **explained that I will be using the word queer!!!**
- **explained that it was the original author's idea to exclude binary trans people, not mine**

In this notebook, I will ____________, including `python` code.

## Table of Contents

- [Data Preparation](##Data-Preparation)
  - [Imports](###Imports)
  - [Renaming Columns](###Renaming-Columns) 
  - [Feature Selection, Part 1](###Feature-Selection-Part-1)
  - [Missing Values and Imputation](###Missing-Values-and-Imputation)
  - [Feature Engineering](###Feature-Engineering)
  - [Feature Selection, Part 2](###Feature-Selection-Part-2)
- [Notebook Summary](##Notebook-Summary)  

## Data Preparation

In the cells below, I have reported the steps I took to prepare the dataset for modeling.  Readers should recall from the Introduction and Methods notebook that none of the code will run unless they download the dataset from DSDR's archive, and import it when prompted below.  **Anyone wishing to reproduce my work must first obtain a copy of the dataset from DSDR.**

Where neccessary, I have provided commentary on the code below.  Otherwise, I trust that readers will the code self-explanatory.

### Imports

In [2]:
# Import modules
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from emilys_functions import my_date, autoplots

In [3]:
# Settings preferences
pd.set_option('display.max_rows', None)
pd.options.mode.chained_assignment = None 

While preparing this report on my local computer, I used the absolute path in the cell below, which has since been commented out, to import my copy of the dataset.  Any readers wishing to follow along should uncomment the code in the cell after that, and modify it to point to the location of the file on their own computer.  I named the resulting dataframe `meyer`, in honor of Ilan H. Meyer, the author of the original study.  Readers should note that if they change that name in their import cell, they will also have to change it throughout the code.

In [4]:
# Import the data - mine
meyer = pd.read_csv(
  '../../potential_datasets/2024_05_23_download_ICPSR_Meyer_2023_generations_data_attempt_2/ICPSR_37166/DS0007/37166-0007-Data.tsv', 
  sep = '\t', low_memory=False, na_values = ' ') # Many thanks to ibrahim rupawala for highlighting the na_values argument
  # https://stackoverflow.com/questions/13445241/replacing-blank-values-white-space-with-nan-in-pandas/47105408#47105408

In [5]:
# Import the data - yours
# meyer = pd.read_csv('your_path/37166-0007-Data.tsv', 
#   sep = '\t', low_memory=False, na_values = ' ')

In [6]:
# Ours
meyer.shape

(1518, 1329)

### Renaming Columns

The variable names in the original dataset are in all uppercase letters.  Per `python` norms (and, conveniently, my own preference), I converted them to all lowercase letters.  

In [7]:
# Rename the columns
meyer.columns = [c.lower() for c in list(meyer.columns)]

### Feature Selection, Part 1

Meyer's original study was longitudinal, so the dataset he provided included responses from the participants at up to three different times of measurement, referred to as waves.  Many authors have done excellent analyses on this time-series element.  However, as is often the case, Meyer's study suffered from attrition effects, and only 616 participants responded to all three rounds of data collection.  In the interest of preserving a larger sample, I chose to cross-sectionally investigate the wave 1 responses only, for an initial sample size of 1518.  Therefore, I dropped the columns pertaining to wave 2 or 3.

In [8]:
# Display the n per wave; note that 4 = all 3 waves
meyer['waveparticipated'].value_counts(dropna = False).sort_index()

waveparticipated
1    533
2    278
3     91
4    616
Name: count, dtype: int64

In [9]:
# Eliminate W2 and W3 variables from the list
cols = list(meyer.columns)
w1_cols = [c for c in cols if c[:2]!='w2']
w1_cols = [c for c in w1_cols if c[:2]!='w3']
len(w1_cols)

# Drop columns
meyer = meyer[w1_cols]
meyer.shape

(1518, 505)

Additionally, as described on page 17 of *37166-Documentation-methodology.pdf*, several participants identified as heterosexual on the survey instrument, despite indicating that they were LGB on the screening questionaire.  Although these people likely have a more complicated identity than either of those options can fully capture, I opted to err on the side of caution and exclude them from my analyses.

In [10]:
# Exclude anyone marking themselves as straight/heterosexual
meyer = meyer[meyer['w1sexualid']!=1]
meyer.shape

(1507, 505)

Next, I sought to eliminate the columns corresponding to items in scales, in favor of using the scale scores themselves, which were kindly already calculated and included as their own columns.

In [11]:
# Make lists of the items that comprise scales
soc_supp_items = ['w1q164a', 'w1q164b', 'w1q164c', 'w1q164d', 'w1q164e', 
  'w1q164f', 'w1q164g', 'w1q164h', 'w1q164i', 'w1q164j', 'w1q164k', 'w1q164l']

ace_items = ['w1q151', 'w1q152', 'w1q153', 'w1q154', 'w1q155', 
  'w1q156', 'w1q157', 'w1q158', 'w1q159', 'w1q160', 'w1q161']

childhd_gnc_items = ['w1q147', 'w1q148', 'w1q149', 'w1q150']

daily_discr_items = ['w1q144a', 'w1q144b', 'w1q144c','w1q144d', 
  'w1q144e', 'w1q144f', 'w1q144g', 'w1q144h', 'w1q144i']

int_homo_items = ['w1q128', 'w1q129', 'w1q130', 'w1q131', 'w1q132']

felt_stigma_items = ['w1q125', 'w1q126', 'w1q127']

drug_items = ['w1q90', 'w1q91', 'w1q92', 'w1q93', 'w1q94', 
  'w1q95', 'w1q96', 'w1q97', 'w1q98', 'w1q99', 'w1q100']

alc_items = ['w1q85', 'w1q86', 'w1q87']

hc_ster_threat_items = ['w1q60', 'w1q61', 'w1q62', 'w1q63']

comm_conn_items = ['w1q53', 'w1q54', 'w1q55', 'w1q56', 'w1q57', 'w1q58', 'w1q59']

lgbis_items = ['w1q40', 'w1q41', 'w1q42', 'w1q43', 'w1q44']

meim_items = ['w1q21', 'w1q22', 'w1q23', 'w1q24', 'w1q25', 'w1q26']

swl_items = ['w1q186', 'w1q187', 'w1q188', 'w1q189', 'w1q190']

swb_items = ['w1q04', 'w1q05', 'w1q06', 'w1q07', 'w1q08', 'w1q09', 'w1q10', 
  'w1q11', 'w1q12', 'w1q13', 'w1q14', 'w1q15', 'w1q16', 'w1q17', 'w1q18']

# This one is the y variable, so I'm not dropping them just yet
ment_dis_items = ['w1q77a', 'w1q77b', 'w1q77c', 'w1q77d', 'w1q77e', 'w1q77f']

# I will drop the others though
drop_cols = (soc_supp_items + ace_items + childhd_gnc_items + daily_discr_items + 
  int_homo_items + felt_stigma_items + drug_items + alc_items + comm_conn_items + 
  hc_ster_threat_items + lgbis_items + meim_items + swl_items + swb_items)
len(drop_cols)

100

In [12]:
# Drop the scale component columns
meyer.drop(columns = drop_cols, inplace = True)
meyer.shape

(1507, 405)

In [13]:
# These are the scales - more on them later
combo_features = ['w1socialwb', 'w1socialwb_i', 'w1lifesat', 'w1lifesat_i', 'w1meim', 'w1meim_i', 
  'w1idcentral', 'w1idcentral_i', 'w1connectedness', 'w1connectedness_i', 'w1hcthreat', 
  'w1hcthreat_i', 'w1kessler6', 'w1kessler6_i', 'w1auditc', 'w1auditc_i', 'w1dudit', 'w1dudit_i', 
  'w1feltstigma', 'w1feltstigma_i', 'w1internalized', 'w1internalized_i', 'w1everyday', 
  'w1everyday_i', 'w1childgnc', 'w1childgnc_i', 'w1ace', 'w1ace_i', 'w1ace_emo', 'w1ace_emo_i', 
  'w1ace_inc', 'w1ace_inc_i', 'w1ace_ipv', 'w1ace_ipv_i', 'w1ace_men', 'w1ace_men_i', 'w1ace_phy', 
  'w1ace_phy_i', 'w1ace_sep', 'w1ace_sep_i', 'w1ace_sex', 'w1ace_sex_i', 'w1ace_sub', 'w1ace_sub_i', 
  'w1socsupport', 'w1socsupport_fam', 'w1socsupport_fam_i', 'w1socsupport_fr', 'w1socsupport_fr_i', 
  'w1socsupport_i', 'w1socsupport_so', 'w1socsupport_so_i']

Keen-eyed readers may have noticed a pattern in the list above: each variable name appears twice, with `'_i'` added to the end of the second occurrence.  This is because Meyer and his team provided two versions of the computed column for each scale - one for the original values, and one in which missing values have been imputed.  I discuss imputation - these, and my own - in the next section.

### Missing Values and Imputation

Because this dataset only has about 1500 rows, I felt it important to preserve as many of them as possible.  Out of this necessity, I opted to allow a larger proportion of imputed data than I would b inclined to do if I had rows to spare.  The implications of this decision will be discussed at further length in the *Limitations* section, but suffice to say it will require me to interpret my models cautiously and conservatively.

Fortunately, however, Meyer and his team used thorough, theoretically-sound methods for their imputations, inpsiring confidence in me regarding the quality of the imputed data.  Where possible, they filled missing values with logical inferences based on that participant's other responses.  Where that was not possible, they predicted the missing value based on a regression of the participant's other responses, then imputed a randomly selected value from the 5 extant values nearest to that prediction.  See page 30 of *37166-Documentation-methodology.pdf* for more details.

# **MAYBE TAKE THIS OUT???  I CAN'T TELL IF IT'S CONVINCING**  vv

For a quick-and-dirty test of the facial validity of these imputations, I compared the value counts of `w1ace_ipv` (before imputation) and `w1ace_ipv_i` (after imputation).  This variable corresponds to questions from the Adverse Childhood Experiences (ACE) scale regarding intimate partner violence (IPV) in their home when they were a child.  IPV is notoriously underreported, and not surprisingly, this variable had one of the highest rates of missing data.

In [14]:
print(meyer['w1ace_ipv'].value_counts(dropna = False))
print(meyer['w1ace_ipv_i'].value_counts(dropna = False))

w1ace_ipv
0.0    945
1.0    426
NaN    136
Name: count, dtype: int64
w1ace_ipv_i
0    1018
1     489
Name: count, dtype: int64


In [15]:
1010-940, 484-424

(70, 60)

In [16]:
len(meyer[(meyer['w1ace_ipv']==1) & (meyer['w1ace_sex'].isna())])

19

In [17]:
len(meyer[(meyer['w1ace_ipv'].isna()) & (meyer['w1ace_sex']==1)])

50

In [18]:
meyer[(meyer['w1ace_ipv'].isna()) & (meyer['w1ace_sex']==1)]['w1ace_ipv_i'].value_counts(dropna = False)

w1ace_ipv_i
1    27
0    23
Name: count, dtype: int64

In [19]:
meyer[(meyer['w1ace_ipv'].isna()) & (meyer['w1ace_ipv_i']==0)]['w1ace_sex'].value_counts(dropna = False, normalize = True)

w1ace_sex
0.0    0.438356
1.0    0.315068
NaN    0.246575
Name: proportion, dtype: float64

In [20]:
meyer[(meyer['w1ace_ipv'].isna()) & (meyer['w1ace_ipv_i']==1)]['w1ace_sex'].value_counts(dropna = False, normalize = True)

w1ace_sex
1.0    0.428571
0.0    0.380952
NaN    0.190476
Name: proportion, dtype: float64

In [21]:
meyer['w1ace_ipv'].value_counts(dropna = True, normalize = True)

w1ace_ipv
0.0    0.689278
1.0    0.310722
Name: proportion, dtype: float64

In [22]:
len(meyer[(meyer['w1ace_ipv']==' ') & (meyer['w1ace_sex']=='1')])

0

Of the 130 missing values in the original column, 

# ^^ **MAYBE TAKE THIS OUT???  I CAN'T TELL IF IT'S CONVINCING**

Heartened by this level of care - and lacking any better ideas for imputing those values - I was inclined to retain the imputed versions and drop the original columns.  Before doing so, however, I wanted to assess how many values had originally been missing.  Features for which a high proportion of values had to be imputed may not be suitable for modeling, and observations for which a high proportion of the Kessler-6 items had to be imputed may also be inappropriate for use.  

For all of the reasons outlined above, I opted not to immediately drop any features based solely on their original missing proportions, but simply to monitor their performance in the model(s) with an extra watchful eye.  I also decided to keep observations for which only one value out of the 6 items on the Kessler scale had to be imputed, but I did opt to drop any observations with two or more originally missing values among the Kessler items.

In [23]:
# Create a subsetted dataframe to identify any rows to drop
meyer_kessler = meyer[ment_dis_items + ['w1kessler6', 'w1kessler6_i', 'studyid']]
meyer_kessler = meyer_kessler[meyer_kessler['w1kessler6'].isna()]
meyer_kessler

w1q77a  w1q77b  w1q77c  w1q77d  w1q77e  w1q77f  w1kessler6  \
39       NaN     NaN     NaN     NaN     NaN     NaN         NaN   
50       1.0     NaN     NaN     NaN     NaN     NaN         NaN   
96       5.0     5.0     NaN     4.0     3.0     4.0         NaN   
155      5.0     NaN     5.0     4.0     5.0     5.0         NaN   
214      NaN     NaN     NaN     NaN     NaN     NaN         NaN   
227      NaN     NaN     NaN     NaN     NaN     NaN         NaN   
415      NaN     NaN     NaN     NaN     NaN     NaN         NaN   
447      1.0     NaN     3.0     3.0     4.0     2.0         NaN   
457      NaN     NaN     NaN     NaN     NaN     NaN         NaN   
577      NaN     NaN     NaN     NaN     NaN     NaN         NaN   
648      3.0     5.0     4.0     5.0     NaN     5.0         NaN   
677      1.0     1.0     NaN     3.0     3.0     1.0         NaN   
820      4.0     NaN     4.0     3.0     4.0     5.0         NaN   
866      3.0     NaN     4.0     4.0     NaN     NaN         NaN   
943      NaN     NaN     NaN     NaN     NaN     NaN         NaN   
964      2.0     NaN     3.0     NaN     NaN     NaN         NaN   
1090     3.0     5.0     3.0     5.0     NaN     5.0         NaN   
1096     3.0     3.0     2.0     NaN     3.0     2.0         NaN   
1113     NaN     NaN     NaN     NaN     NaN     NaN         NaN   
1159     NaN     NaN     NaN     NaN     NaN     NaN         NaN   
1265     3.0     3.0     3.0     5.0     NaN     5.0         NaN   
1272     2.0     NaN     3.0     4.0     4.0     5.0         NaN   
1344     5.0     NaN     5.0     5.0     5.0     5.0         NaN   
1359     NaN     NaN     NaN     NaN     NaN     NaN         NaN   
1472     3.0     NaN     3.0     2.0     3.0     2.0         NaN   
1505     5.0     3.0     1.0     2.0     NaN     2.0         NaN   

      w1kessler6_i    studyid  
39               1  151876818  
50               9  152304751  
96               4  152640441  
155              3  153039829  
214              0  153390973  
227              9  153535970  
415              8  154944576  
447             16  155106769  
457             12  155163167  
577              2  155814388  
648              4  156224369  
677             20  156345609  
820              5  157670785  
866              6  158132069  
943             12  159311507  
964              7  159770206  
1090             5  162335772  
1096            15  162373382  
1113            16  162548926  
1159             4  162982357  
1265             8  164454095  
1272             9  164556482  
1344             0  167022959  
1359             6  167341434  
1472            15  173409447  
1505            16  175457423

In [24]:
# Extract the index numbers of rows with 2+ NAs

# Get the unique identifier
drop_rows = []
for i in meyer_kessler.index:
  if meyer_kessler.loc[i, ment_dis_items].isna().sum() > 1:
    drop_rows.append(meyer_kessler.loc[i, 'studyid'])

# Get the matching index number from the full dataframe
drop_index = []
for i in drop_rows:
  drop_index.append(meyer[meyer['studyid']==i].index[0])

# Drop those rows
meyer.drop(index = drop_index, inplace = True)
meyer.shape

(1494, 405)

In [25]:
# Reset the index (thanks to pandas documentation for this syntax)
meyer.reset_index(drop = True, inplace = True)
meyer.head(3)

studyid  waveparticipated  w1weight_full  w1weight_orig  \
0  151339842                 4       0.309003       0.282558   
1  151351600                 2       0.931862       0.836694   
2  151396232                 1       0.656089       0.606815   

   w1cumulative_wt_nr1  w1cumulative_wt_nr2  w1cumulative_wt_nr3  \
0           281.512121           368.596299           659.611404   
1           874.104463          1104.824533          2724.529946   
2          1018.123812          1071.729432          1635.443546   

   w1cumulative_wt_sampling  w1weighting_cell_nr1  w1weighting_cell_nr2and3  \
0                368.596299               21313.0                    3112.0   
1               1104.824533               23212.0                    2131.0   
2               1071.729432               22223.0                    2222.0   

   ...  w1socsupport_so_i  gmethod_type_w2  grespondent_date_w2  gsurvey  \
0  ...               4.75              2.0              20958.0      2.0   
1  ...               6.00              1.0              21070.0      NaN   
2  ...               7.00              NaN                  NaN      NaN   

   gmethod_type_w3  gp2  grace grespondent_date_w3 wave3  nopolicecontact  
0              2.0  2.0    1.0             21306.0   1.0              1.0  
1              NaN  NaN    NaN                 NaN   NaN              NaN  
2              NaN  NaN    NaN                 NaN   NaN              NaN  

[3 rows x 405 columns]

With those rows removed, I next checked for missing values in the non-imputed scale score columns.

In [26]:
meyer[combo_features].isna().sum().sort_values(ascending = False)

w1ace                 267
w1ace_ipv             130
w1ace_emo              83
w1ace_sex              69
w1dudit                56
w1ace_phy              54
w1socialwb             51
w1connectedness        39
w1socsupport           37
w1everyday             31
w1meim                 22
w1socsupport_fr        22
w1socsupport_fam       21
w1socsupport_so        20
w1lifesat              19
w1internalized         19
w1hcthreat             16
w1ace_inc              16
w1ace_sub              15
w1ace_men              14
w1idcentral            14
w1kessler6             13
w1ace_sep              10
w1feltstigma            8
w1auditc                6
w1ace_men_i             0
w1socsupport_fam_i      0
w1ace_sub_i             0
w1ace_sex_i             0
w1ace_sep_i             0
w1socsupport_fr_i       0
w1ace_phy_i             0
w1socsupport_i          0
w1lifesat_i             0
w1idcentral_i           0
w1ace_emo_i             0
w1ace_ipv_i             0
w1ace_inc_i             0
w1meim_i    

As mentioned above, I was impressed enough by the original author's imputation methods to not drop any of these features over their missing values alone.  However, quite a few columns had troublingly high numbers of missing values.  Those with approximately 1% missing data or more are displayed below.

|Variable Name| Missing Values|Variable Name| Missing Values|Variable Name| Missing Values|
|--|--|--|--|--|--|
|w1ace        |267 |w1connectedness  | 39 |w1lifesat       | 19 |
|w1ace_ipv    |130 |w1socsupport     | 37 |w1internalized  | 19 |
|w1ace_emo    | 83 |w1everyday       | 31 |w1hcthreat      | 16 |
|w1ace_sex    | 69 |w1meim           | 22 |w1ace_inc       | 16 |
|w1dudit      | 56 |w1socsupport_fr  | 22 |w1ace_sub       | 15 |
|w1ace_phy    | 54 |w1socsupport_fam | 21 |w1ace_men       | 14 |
|w1socialwb   | 51 |w1socsupport_so  | 20 |w1idcentral     | 14 |

As can be seen above, many of my most theory-backed potential predictors have large numbers of missing values.  The implications of this will be discussed in the Limitations section, but in the interest of not sacrificing too much data, I went forward with my plan to drop these original columns and use the imputed versions.

In [27]:
# Divide that combination list between imputed and not
combo_imputed = [c for c in combo_features if c[-2:]=='_i']
combo_not_imputed = [c for c in combo_features if c[-2:]!='_i']

# Drop the ones that aren't
meyer.drop(columns = combo_not_imputed, inplace = True)
meyer.shape

(1494, 379)

Before investigating any further missing values, I went through the documentation to identify and drop any columns that were not pertinent to my problem statement.

In [28]:
# Identify excess columns
drop_cols_2 = ['w1q165', 'w1q27', 'w1q28', 'w1q20_1', 'w1q20_2', 'w1q20_3', 
  'w1q20_4', 'w1q20_5', 'w1q20_6', 'w1q20_7', 'w1q29', 'w1q29_t_verb', 'gruca', 
  'gruca_i', 'gurban', 'gzipstate',  'gzipcode', 'w1hinc', 'w1poverty', 
  'w1povertycat', 'w1q133', 'w1q133_1', 'w1q133_2', 'w1q133_3', 
  'gmsaname', 'gmethod_type', 'gmethod_type_w2', 'gmethod_type_w3', 
  'w1cumulative_wt_nr1', 'w1cumulative_wt_nr2', 'w1cumulative_wt_nr3', 
  'w1cumulative_wt_sampling', 'w1weighting_cell_nr1', 'w1weight_orig', 
  'w1weighting_cell_nr2and3', 'w1frame_wt', 'w1q31a', 'w1q31b', 'w1q31c', 
  'w1q31d', 'w1q66_1', 'w1q66_2', 'w1q66_3', 'w1q66_4', 'w1q66_5', 
  'w1q66_t_verb', 'w1q67', 'w1q68_1', 'w1q68_2', 'w1q68_3', 'w1q70', 'w1q71', 
  'w1q73', 'w1q74_1', 'w1q74_2', 'w1q74_3', 'w1q74_4', 'w1q74_7', 'w1q74_8', 
  'w1q74_9', 'w1q74_12', 'w1q74_13', 'w1q74_15', 'w1q74_16', 'w1q74_19', 
  'w1q80', 'w1q81', 'w1q82', 'w1q83', 'w1q84', 'w1q88', 'w1q118', "gp1"
  "w1q20_t_verb", "w1q39_7", "w1q39_12", "w1q39_t_verb", "w1q39_6", 
  "w1q39_5", "w1q39_2", "w1q39_3", "w1q39_4", "w1q39_1", "w1q39_9", "w1q39_8", 
  "w1q39_10", "w1q39_11", "gemployment2010", "gmilesaway"]
meyer.drop(columns = drop_cols_2, inplace = True)
meyer.shape

(1494, 307)

### Feature Engineering

- creating new columns
- polynomial terms
- standardization

### Feature Selection, Part 2

## Notebook Summary

In this notebook, I have __________.  

In the following notebook, I will ___________.

Any readers who are following along or attempting to reproduce my work should uncomment the cell below to save a copy of the dataframe as it exists now.  A cell is provided at the top of the next notebook in which to import that copy.

In [29]:
# Save a copy of the dataframe to use in the next notebook
meyer.to_csv('../02_data/df_after_data_preparation.csv', index = False)